File: 040_combination_sum_II.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        Arrays.sort(candidates);
        backtracking(result, candidates, new ArrayList<>(), target, 0);
        return result;
    }
    
    private void backtracking(List<List<Integer>> result, int[] candidates, List<Integer> tempList, int remain, int start){
        if (remain < 0) return;
        else if(remain == 0) result.add(new ArrayList<>(tempList));
        else{
            for (int i = start; i < candidates.length; i++){
                if(i > start && candidates[i] == candidates[i-1]) continue; // skip duplicates
                tempList.add(candidates[i]);
                backtracking(result, candidates, tempList, remain - candidates[i], i + 1);
                tempList.remove(tempList.size() - 1);
            }
        }
    }
}

File: python/040_combination_sum_II.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        def backtrace(cand: List[int], res: List[List[int]], temp: List[int], curr: int, target: int, start: int):
            if curr > target:
                return
            if curr == target:
                res.append(list(temp))
            for i in range(start, len(cand)):
                if i > start and cand[i] == cand[i - 1]: continue
                temp.append(cand[i])
                backtrace(cand, res, temp, curr + cand[i], target, i + 1)
                temp.pop()

        candidates.sort()
        res = []
        backtrace(candidates, res, [], 0, target, 0)
        return res
